### ISSUE TITLE
missing data

### ISSUE THREAD

**Issue by yuchenhou** [yuchenhou/elephant#29]
# Background

Typically, we use machine learning to solve problems with data sets like this:

| x_1 | x_2 | x_3 | y (target) |
| --- | --- | --- | --- |
| 2.5 | 3 | 5.9 | 4.2 |
| 2 | 3.2 | 9 | 43.5 |
| ... | ... | ... | ... |
| 6.2 | 3 | 439 | 74.5 |
| 22.5 | 34 | 5.39 | ? |
| 232 | 36.2 | 9 | ? |
| ... | ... | ... | ... |
| 936.2 | 63 | 329 | ? |

That is, samples in test set have target attributes with unknown values; all other attributes have known values. The model learns on the train set and predicts on the test set.
# Motivation

However, in my internship, most of problems have data sets like this

| x_1 | x_2 | x_3 | x_4 |
| --- | --- | --- | --- |
| 2.5 | 3 | 5.9 | 4.2 |
| 2 | ? | ? | 43.5 |
| ... | ... | ... | ... |
| 6.2 | 3 | 439 | 74.5 |
| 22.5 | 34 | ? | 5.39 |
| ? | 232 | 36.2 | 9 |
| ... | ... | ... | ... |
| 936.2 | 63 | ? | ? |

That is, every sample can have an arbitrary subset attributes with unknown values. It turned out that this is an existing and important problem called missing data. A quick search reveals a Wikipedia article on [Missing data](https://en.wikipedia.org/wiki/Missing_data), which mentions "Model based techniques, often using graphs...". I can do some research on this and find out the current progress of neural net approaches on this, if any.
